rasa/engine/caching.py

```diff
@@ -0,0 +1,373 @@
+            model_storage: Required for caching `Cacheable` instances. E.g. `Resource`s
+                use that to copy data from the model storage to the cache.
+        """
+        if self._is_disabled():
+            return
+
+        cache_dir, output_type = None, None
+        if isinstance(output, Cacheable):
+            cache_dir, output_type = self._cache_output_to_disk(output, model_storage)
+
+        with self._sessionmaker.begin() as session:
+            cache_entry = self.CacheEntry(
+                fingerprint_key=fingerprint_key,
+                output_fingerprint_key=output_fingerprint,
+                last_used=datetime.utcnow(),
+                rasa_version=rasa.__version__,
+                result_location=cache_dir,
+                result_type=output_type,
+            )
+            session.add(cache_entry)
+
+    def _is_disabled(self) -> bool:
+        return self._max_cache_size == 0.0
+
+    def _cache_output_to_disk(
+        self, output: Cacheable, model_storage: ModelStorage
+    ) -> Tuple[Optional[Text], Optional[Text]]:
+        with tempfile.TemporaryDirectory() as temp_dir:
+            try:
+                output.to_cache(Path(temp_dir), model_storage)
+
+                logger.debug(f"Caching output of type '{type(output)}' succeeded.")
+            except Exception as e:
+                logger.error(
+                    f"Caching output of type '{type(output)}' failed with the "
+                    f"following error:\n{e}"
+                )
+                return None, None
+
+            output_size = _directory_size_in_mb(temp_dir)
+            if output_size > self._max_cache_size:
+                logger.debug(
+                    f"Caching result of type '{type(output)}' was skipped "
+                    f"because it exceeds the maximum cache size of "
+                    f"{self._max_cache_size} MiB."
+                )
+                return None, None
+
+            while (
+                _directory_size_in_mb(self._cache_location) + output_size
+                > self._max_cache_size
+            ):
+                self._drop_least_recently_used_item()
+
+            output_type = rasa.shared.utils.common.module_path_from_instance(output)
+            cache_path = shutil.move(temp_dir, self._cache_location)
+
+            return cache_path, output_type
+
+    def _drop_least_recently_used_item(self) -> None:
+        with self._sessionmaker.begin() as session:
+            query_for_least_recently_used_entry = sa.select(self.CacheEntry).order_by(
+                self.CacheEntry.last_used.asc()
+            )
+            oldest_cache_item = (
+                session.execute(query_for_least_recently_used_entry).scalars().first()
+            )
+
+            # TODO: Can't happen
+            if not oldest_cache_item:
+                return
+
+            self._delete_cached_result(oldest_cache_item)
+            delete_query = sa.delete(self.CacheEntry).where(
```

Comment:
Is this duplicate functionality of `_delete_incompatible_entries_from_cache`?

Reply:
The other one deletes using a list of fingerprint keys where this one uses a single one. We could indeed deduplicate it but I'm wondering whether this doesn't decrease the readability

Reply:
👍  fine as is.